lambdas/services/token_handler_ssm_service.py

```diff
@@ -14,68 +20,116 @@ def get_smartcard_role_codes(self) -> list[str]:
         logger.info("starting ssm request to retrieve required smartcard role codes")
         params = self.get_ssm_parameters(
             [
-                "/auth/smartcard/role/gp_admin",
-                "/auth/smartcard/role/gp_clinical",
-                "/auth/smartcard/role/pcse",
+                GP_ADMIN_USER_ROLE_CODE,
+                GP_CLINICAL_USER_ROLE_CODE,
+                PCSE_USER_ROLE_CODE,
             ]
         )
 
         response = [
-            params["/auth/smartcard/role/gp_admin"],
-            params["/auth/smartcard/role/gp_clinical"],
-            params["/auth/smartcard/role/pcse"],
+            params.get(GP_ADMIN_USER_ROLE_CODE),
+            params.get(GP_CLINICAL_USER_ROLE_CODE),
+            params.get(PCSE_USER_ROLE_CODE),
         ]
 
+        if None in response:
+            logger.error(
+                "SSM parameter values for GP admin/clinical or PSCE roles may not exist",
+                {"Result": "Unsuccessful login"},
+            )
+            raise LoginException(
+                500, "Failed to find SSM parameter value for user role"
+            )
+
         return response
 
-    def get_smartcard_role_gp_admin(self) -> str:
+    def get_smartcard_role_gp_admin(self) -> list[str]:
         logger.info(
             "starting ssm request to retrieve required smartcard role code gp admin"
         )
-        params = self.get_ssm_parameters(["/auth/smartcard/role/gp_admin"])
-
-        response = params["/auth/smartcard/role/gp_admin"]
+        params = self.get_ssm_parameters([GP_ADMIN_USER_ROLE_CODE])
+        values = params.get(GP_ADMIN_USER_ROLE_CODE)
+
+        if values is None:
+            logger.error(
+                "SSM parameter values for GP admin role may not exist",
+                {"Result": "Unsuccessful login"},
+            )
+            raise LoginException(
```

Comment:
Tempted to say this should not be an explicit LoginException, this is more a SSMException

Reply:
I considered this, the reason it's a LoginException is it's an extension of the LambdaException class and not a regular exception so will be caught as part of the lambda exception handler decorator. Also, since this token ssm service is only used by the token handler, made sense to make it a LoginException instead. The alternative would be creating a new SSMException, passing it up to the login service, catching it there and then throwing the LoginException up to the handler so thought it'd be easier to cut out the middleman, but can do either really

Reply:
Okay makes sense, let's monitor this one going forwards. We may need a SSMException in other Lambdas going forwards, if we do, let's hot swap it then.